### PR TITLE
UPnP support

### PR DESCRIPTION
**Pull request by asdacap**
Add automatic port forwarding via UPnP.

## Changes:
- Add a plugin that set up portforwarding via Open.NAT library.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [X] No

**Comments about testing , should you have some** (optional)

- Confirm port forwarding.
- Confirm if I disable UPnP on my router, the error message shows up.
- Confirm it will re-add the entry after expiry interval.

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+{
+    public string Name => "UPnP";
+    public string Description => "Automatic port forwarding with UPnP";
+    public string Author => "Nethermind";
+
+    // Routers tend to clean mapping, so we need to periodically
+    private readonly TimeSpan ExpirationRate = TimeSpan.FromMinutes(10);
+    private CancellationToken _cancellationToken;
+    private INetworkConfig _networkConfig;
+    private ILogger _logger;
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:23]: resolve warning?

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+public class UPnPPlugin: INethermindPlugin
+{
+    public string Name => "UPnP";
+    public string Description => "Automatic port forwarding with UPnP";
+    public string Author => "Nethermind";
+
+    // Routers tend to clean mapping, so we need to periodically
+    private readonly TimeSpan ExpirationRate = TimeSpan.FromMinutes(10);
+    private CancellationToken _cancellationToken;
+    private INetworkConfig _networkConfig;
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:22]: resolve warning?

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            return;
+        }
+
+        await device.CreatePortMapAsync(new Mapping(
+            Protocol.Tcp,
+            _networkConfig.P2PPort,
+            _networkConfig.P2PPort,
+            ExpirationRate.Milliseconds + 10000,
+            "Nethermind P2P"));
+        await device.CreatePortMapAsync(new Mapping(
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:87]: maybe check if they are not the same before calling a new mapping?

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    // Routers tend to clean mapping, so we need to periodically
+    private readonly TimeSpan ExpirationRate = TimeSpan.FromMinutes(10);
+    private CancellationToken _cancellationToken;
+    private INetworkConfig _networkConfig;
+    private ILogger _logger;
+
+    public Task Init(INethermindApi api)
+    {
+        CancellationTokenSource cts = new();
+        api.DisposeStack.Push(new Reactive.AnonymousDisposable(() => cts.Cancel()));
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:28]: cant we just call this on the disposal of the plugin instead of this line?

**Review by smartprogrammer93 — APPROVED** [review]
Needed feature, thank you. there is also spacing issues. so maybe run dotnet format.

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:83]: They have different protocol, so might as well be separate anyway.

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs]: Done

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            return;
+        }
+
+        await device.CreatePortMapAsync(new Mapping(
+            Protocol.Tcp,
+            _networkConfig.P2PPort,
+            _networkConfig.P2PPort,
+            ExpirationRate.Milliseconds + 10000,
+            "Nethermind P2P"));
+        await device.CreatePortMapAsync(new Mapping(
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:83]: This will be continuously called in a loop?

Comment (asdacap) [reply src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:83]: There is a `WaitOnAsync` at higher level.

File: src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            {
+                if (_logger.IsWarn) _logger.Error("Unable to setup UPnP mapping.", exception);
+            }
+
+            if (await _cancellationTokenSource.Token.WaitHandle.WaitOneAsync(ExpirationRate, CancellationToken.None))
+            {
+                break;
+            }
+        } while (true);
+    }
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.UPnP.Plugin/UPnPPlugin.cs:56]: Yes, see comment for expiration rate.

**Pull request merged**